### ISSUE TITLE
syntastic and :mksession

### ISSUE THREAD

**Issue by cfi** [scrooloose/syntastic#1596]
Question (short):
Is there any way to combine loading a session _and_ having `syntastic_check_on_open=1`?

Long version:

When saving sessions, I followed the advice in chapter 6.4 of the docs:

> 6.4 Saving Vim sessions *syntastic-sessions*
> If you use |:mksession| to save Vim sessions you should probably make sure to remove option "blank" from 'sessionoptions' ... This will prevent |:mksession| from saving |syntastic-error-window| as empty quickfix windows.

So I have in my .vimrc:

    set sessionoptions=curdir,folds,resize,tabpages,winpos,winsize 

Then I use `:vsplit` to create three columns of views and load a different file into each. After loading each file, syntastic opens its little error view near the bottom of the respective vsplit.
Now, `:mksession session.vim` saves the session.

Closing vim and reloading with `vim -S session` does reload the three files into the three vsplits. But the splits done by syntastic mess up the layout: The error view of the first file does not stretch only beneath the first file, but stretches across the full width of the window. The second and third error views are displayed beneath the second file (so near the bottom of the second vsplit).

Probably this is because the error views are opened while the session is loaded, aren't they? Imho they should open after all session load activities are done and settled.

The easy fix is to include

    let g:syntastic_check_on_open = 0

in the `~/.vimrc`.

Is there any way to combine loading a session _and_ having `syntastic_check_on_open=1`?

**Comment by cfi** [comment]
It's unrelated, but just to double check I also saved sessions with

    set sessionoptions=curdir

only, and it is still the same behaviour. I believe the problem really is with the order of actions during session loading. And I do not know any of the vim internals, and if it is possible to hook into that order and react differently than on ordinary file loads.

Session files are wrapped with a

    let SessionLoad = 1
    ...
    unlet SessionLoad

so maybe we could postpone the syntax check on open (if enabled) to _after_ the `unlet SessionLoad`?

**Comment by lcd047** [comment]
> Is there any way to combine loading a session _and_ having `syntastic_check_on_open=1`?

Well, you could say syntastic itself works as intended, it's your expectation that is [unrealistic](https://github.com/scrooloose/syntastic/pull/1184#issuecomment-55456628).

Sadly, the problem you describe can't be fixed without breaking other things.  Vim can't detect when a session loading has started, it can only find out after the fact, when loading has finished, and that's too late for doing anything.  Sorry about that.

A workaround for you might be to leave `check_on_open` unset, and set it only after startup:
```vim
let g:syntastic_check_on_open = 0
autocmd VimEnter * let g:syntastic_check_on_open = 1
```
The downside is that files opened with `vim some_file` when no session is involved won't be checked on open, either. _shrug_

**Comment by cfi** [comment]
Oh. I'm surprised. I thought the syntastic buffers are openend during the session loading because vim actually calls a syntastic callback during loading a file. Not knowing vim's API the behaviour looked as if syntastic registered a hook with file loading.

The second assumption was that this callback does in fact see the `SessionLoad` variable that is defined during session load.

It seems one or both of these assumptions are wrong then.

Anyway, thanks for syntastic and your clarification on this matter!
I can relate to your rant in the referenced "[unrealistic](https://github.com/scrooloose/syntastic/pull/1184#issuecomment-55456628)" pull request about having to deal with circuitious ways to get to the relevant information without having proper APIs. Someone has to get their hands dirty with that crap, if there are no gloves and shovels. ;-)

I can live with the workaround.

**Comment by lcd047** [comment]
Like I said: not without breaking other things.
